### ISSUE TITLE
Unexpected end of input

### ISSUE THREAD

**Issue by klarkent** [valery-barysok/session-file-store#21]
I'm having this exception and sometimes this makes express return error 500:
SyntaxError: Unexpected end of input
    at Object.parse (native)
    at /home/salvo/.local/lib/node_modules/session-file-store/node_modules/fs-extra/node_modules/jsonfile/index.js:14:18
    at fs.js:334:14
    at /home/salvo/.local/lib/node_modules/session-file-store/node_modules/fs-extra/node_modules/graceful-fs/graceful-fs.js:42:10
    at FSReqWrap.oncomplete (fs.js:95:15)

**Comment by jabez128** [comment]
you can try change the `retry` option to 50 or more :)

**Comment by brunobottazzini** [comment]
I have the same error here...

@jabez128, I have tried setting retry by running:

        var session = require('express-session');
        var SessionStore = require('session-file-store')(session);

        var SessionOptions = { 
            retries: 50
        };

        app.use(session({
            genid: function(req) { 
                // uuid for guest session ids
                var id = uuid.v4();
                console.log("New session: " + id);
                return id;
            },
            name: "session",
            secret: 'THE_SECRET',
            resave: true,
            saveUninitialized: true,
            store: new SessionStore(SessionOptions)
        }));

Is this correct ? 

I have seen on README.md that it was *RETRIES* and no *RETRY*. 

In any case, I have tested it with both and still I got the error after the fifth attempt. 

It looks like it ignores my options

**Comment by valery-barysok** [comment]
strange, what version do you use?

**Comment by valery-barysok** [comment]
see my test as example https://github.com/valery-barysok/session-file-store/blob/master/test/helpers.js#L68

**Comment by brunobottazzini** [comment]
@valery-barysok, I am running the last one 0.0.20

**Comment by brunobottazzini** [comment]
@valery-barysok thanks!

This is odd, I tried setting the option path like this:

        var SessionOptions = { 
            retries: 50,
            path: "/tmp/sessions"
        };

And it still creates in the default path.

I am wondering what I'm doing wrong if the tests pass successfully...

**Comment by valery-barysok** [comment]
i just tested it on windows and it works. Complete example here: https://gist.github.com/valery-barysok/bed7855e037781b47e79

**Comment by brunobottazzini** [comment]
@valery-barysok I figure it out! 

It was a silly mistake (File permission) now it work smoothly :+1: 

Thank you very much :)

**Comment by valery-barysok** [comment]
:ship:

**Comment by brunobottazzini** [comment]
After some tests this solution proposed works quite fine. I'd like to understand better this issue.

The following message occurs frequently for me:

    [session-file-store] will retry, error on last attempt: SyntaxError: Unexpected end of input

I think setting a bigger retry is a good workaround but, shouldn't session-file-store work without this error occurring ?

Do you know why this occurs ? Maybe I am doing something in my code that can provoke this behavior?

Many thanks

**Comment by valery-barysok** [comment]
in nutshell, it is ok. )) i gonna to add details descriptions for such questions in readme file and recommendations. please, wait for full answer in the near future.

**Comment by brunobottazzini** [comment]
Alright ! Thank you very much :)

**Comment by appernetic** [comment]
I also sometimes has this problem in my express app: 

[session-file-store] will retry, error on last attempt: SyntaxError: /Users/SuperStrength/Documents/Devserver/sessions/-HcNAeJpbl7WFD6bo6g0AC5u6A4oGtfy.json: Unexpected end of input

```
app.use(session({
      store: new FileStore({path: __dirname + '/sessions'}),
      secret: 'My very secret secret',
      resave: false,
      retries: 50,
      saveUninitialized: true
  }));
```
It will not allow more requests until I restart the node server. I have just added retries: 50 to see if it will improve.

**Comment by appernetic** [comment]
No luck, still error and a web app that is waiting for a response.

[session-file-store] will retry, error on last attempt: SyntaxError: /Users/SuperStrength/Documents/Devserver/sessions/Hm13_U9gEXhAVD3Ky8mQqaFGDg8kOQmP.json: Unexpected end of input

**Comment by appernetic** [comment]
Now I tried:

```
app.use(session({
    genid: function(req) {
        // uuid for guest session ids
        var id = uuid.v4();
        console.log("New session: " + id);
        return id;
    },
    name: "session",
    store: new FileStore({
      retries: 50,
      path: __dirname + '/sessions'
    }),
    secret: 'keyboard cat',
    resave: true,
    saveUninitialized: true
  })
);
```

And this time I get more error:

First I get lots of:
```
[session-file-store] will retry, error on last attempt: Error: ENOENT, open '/Users/SuperStrength/Documents/Devserver/sessions/007fb3a7-ea59-4a59-8d05-f72fdc0b16df.json'
````
Then I also get a few Syntax Error after the log message New session: 6e607dce-6f8c-4ca3-b9af-16749e9c9015:

```
[session-file-store] will retry, error on last attempt: SyntaxError: /Users/SuperStrength/Documents/Devserver/sessions/6e607dce-6f8c-4ca3-b9af-16749e9c9015.json: Unexpected end of input
```

**Comment by appernetic** [comment]
Does not work, a restart of express generates 50 or so SyntaxError and the web app hangs.

**Comment by appernetic** [comment]
Sorry, I found out that it was a inherited module from one of my old blueprints. I do not need state in my server so I just removed it.

**Comment by valery-barysok** [comment]
hey @appernetic, i did not clear understand if it was issue related to session-file-store?

**Comment by appernetic** [comment]
It is still a session-file-store issue that I was not able to solve, but I have worked around it by not using it because I do not need it.

**Comment by purnesh** [comment]
I had this issue and had to switch to 'connect-mongodb-session' and store sessions in a temporary database instead.

**Comment by AGuismo** [comment]
Hello, I also have this issue nearly all the time:
[session-file-store] will retry, error on last attempt: SyntaxError: /var/www/mysite/sessions/AP3E_LBVzH59PmPQsWB8szfRlRF0Csih.json: Unexpected end of input

I had to downgrade to 0.0.22, the last version crashed my server all the time. I use the latest version of express and express-session.
Those problem happened mostly when I use the session (of course) with "req.session.anyvariable" but mostly when I create a new session after had previously called "req.session.destroy".

Here is how I use this module:

var Session = require('express-session');
var sessionStore = require('session-file-store')(Session);
var session = Session({store: new sessionStore({path: __dirname+'/tmp/sessions', reapInterval: 800}), secret: 'onceuponatime', resave: false, saveUninitialized: false, cookie: { maxAge: 31 * 24 * 60 * 60 * 1000 }});

If you want any additional information dont hesitate.

Thank you =)

**Comment by valery-barysok** [comment]
@AGuismo, thanks. Will try to see what can i do. But FYI *"Unexpected end of input"* is **NOT** really a bug if it can read file through retry cycle. It is related to concurrent read/write operations on the same file. There is no lock file support in node out of the box and more ever not all file system support that.
And it is strange that you get more error with the latest version because It added encryption support for the session file. So I need to investigate it.

**Comment by valery-barysok** [comment]
@AGuismo, i just reviewed changes of 0.0.23 version and made some changes. Can you test if it still crashes your server all the time?

**Comment by AGuismo** [comment]
Hello, sorry for the delay, I reinstalled the 0.0.23 but could not crashed my server this time (I really should have kept the error last time..). As for the 0.0.24 I'm currently using it, so far so good, I'll keep you updated if it crash the server again. When I updated (and when it crashed repeatedly) to 0.0.23 I had some sessions saved whereas now I had none, maybe having sessions created in 0.0.22 during the update to 0.0.23 is the cause of my crashes ? Thanks for responding so fast =)

**Comment by HighTime80s** [comment]
Hello.
> SyntaxError: Unexpected end of input
This problem seems to occur, when the content of session file (e.g. sessions/*.json) is empty.

**Comment by valery-barysok** [comment]
I believe this issue resolved by #31
